rllib/env/single_agent_episode.py

```diff
@@ -295,159 +283,187 @@ def is_done(self) -> bool:
-        for training an agent's modules (`RLModule`), converting from or to
-        `SampleBatch` can be performed by this function and its counterpart
-        `to_sample_batch()`.
-
-        Args:
-            batch: A `SampleBatch` instance. It should contain only a single episode.
+    def get_sample_batch(self) -> SampleBatch:
+        """Converts a `SingleAgentEpisode` into a `SampleBatch`.
 
         Returns:
-            An `SingleAegntEpisode` instance containing the data from `batch`.
+            A SampleBatch containing all of this episode's data.
         """
-        is_done = (
-            batch[SampleBatch.TERMINATEDS][-1] or batch[SampleBatch.TRUNCATEDS][-1]
-        )
-        observations = np.concatenate(
-            [batch[SampleBatch.OBS], batch[SampleBatch.NEXT_OBS][None, -1]]
-        )
-        actions = batch[SampleBatch.ACTIONS]
-        rewards = batch[SampleBatch.REWARDS]
-        # These are the infos after stepping the environment, i.e. without the
-        # initial info.
-        infos = batch[SampleBatch.INFOS]
-        # Concatenate an intiial empty info.
-        infos = np.concatenate([np.array([{}]), infos])
-
-        # TODO (simon): This is very ugly, but right now
-        #  we can only do it according to the exclusion principle.
-        extra_model_output_keys = []
-        for k in batch.keys():
-            if k not in [
-                SampleBatch.EPS_ID,
-                SampleBatch.AGENT_INDEX,
-                SampleBatch.ENV_ID,
-                SampleBatch.AGENT_INDEX,
-                SampleBatch.T,
-                SampleBatch.SEQ_LENS,
-                SampleBatch.OBS,
-                SampleBatch.INFOS,
-                SampleBatch.NEXT_OBS,
-                SampleBatch.ACTIONS,
-                SampleBatch.PREV_ACTIONS,
-                SampleBatch.REWARDS,
-                SampleBatch.PREV_REWARDS,
-                SampleBatch.TERMINATEDS,
-                SampleBatch.TRUNCATEDS,
-                SampleBatch.UNROLL_ID,
-                SampleBatch.DONES,
-                SampleBatch.CUR_OBS,
-            ]:
-                extra_model_output_keys.append(k)
+        return SampleBatch(self.get_data_dict())
 
-        return SingleAgentEpisode(
-            id_=batch[SampleBatch.EPS_ID][0],
-            observations=observations if is_done else observations.tolist(),
-            actions=actions if is_done else actions.tolist(),
-            rewards=rewards if is_done else rewards.tolist(),
-            t_started=batch[SampleBatch.T][0],
-            is_terminated=batch[SampleBatch.TERMINATEDS][-1],
-            is_truncated=batch[SampleBatch.TRUNCATEDS][-1],
-            infos=infos if is_done else infos.tolist(),
-            extra_model_outputs={
-                k: (batch[k] if is_done else batch[k].tolist())
-                for k in extra_model_output_keys
-            },
-        )
+    def get_observations(self, indices: Optional[Union[int, List[int], slice]] = None) -> Any:
```

Comment:
Added these very practical new APIs to get data from the episode in a user friendly fashion.